sw/device/silicon_creator/lib/crypto/rsa_3072/rsa_3072.s

```diff
@@ -438,6 +438,372 @@ modexp_var_3072_f4:
 
   ret
 
+/**
+ * Subtracts the modulus M from A.
+ *
+ * Flags: After this subroutine, the C flag is set to 1 if the subtraction
+ * underflowed.
+ *
+ * This routine runs in variable time.
+ * @param[in]  x16: dmem pointer to first limb of modulus M
+ * @param[in]  [w4:w15]: operand A
+ * @param[in]  w31: all-zero
+ * @param[out] [w16:w27]: result
+ *
+ * clobbered registers: x8 to x12, w2, w3, w16 to w27
+ * clobbered Flag Groups: FG0
+ */
+subtract_modulus_var:
+
+  /* prepare temporary registers */
+  li        x8, 4
+  li        x9, 2
+  li        x10, 3
+  li        x11, 16
+
+  /* copy pointer to modulus */
+  addi      x12, x16, 0
+
+  /* clear flags */
+  bn.add    w31, w31, w31
+
+  /* Subtract M from A */
+  loopi     12, 4
+    /* w2 <= A[i] */
+    bn.movr   x9, x8++
+    /* w3 <= M[i] */
+    bn.lid    x10, 0(x12++)
+    /* w2 <= w2 - w3 */
+    bn.subb   w2, w2, w3
+    /* out[i] <= w2 */
+    bn.movr   x11++, x9
+
+  ret
+
+/**
+ * Doubles a number and reduces modulo M in-place.
+ *
+ *   Returns: C = (A + A) mod M
+ *
+ * Requires that A < M < 2^3072. Writes output to the A buffer in DMEM. This
+ * routine runs in variable time.
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]  x16: dmem pointer to first limb of modulus M
+ * @param[in]  [w4:w15]: operand A
+ * @param[in]  w31: all-zero
+ * @param[out] [w4:w15]: result C
+ *
+ * clobbered registers: x2, x3, x7 to x12, w2 to w27
+ * clobbered Flag Groups: FG0
+ */
+double_mod_var:
+  /* save copy of pointer to modulus */
+  addi      x12, x16, 0
+
+  /* clear flags */
+  bn.add    w31, w31, w31
+
+  /* compute aa=(A + A):
+       [w4:w15] <= (A+A) mod 2^3072 = aa[0:3071]*/
+  li        x9, 2
+  li        x10, 4
+  loopi     12, 3
+    /* w2 <= a[i] */
+    bn.movr   x9, x10
+    /* w2 <= w2 + w2 */
+    bn.addc   w2, w2, w2
+    /* aa[i] <= w2 */
+    bn.movr   x10++, x9
+
+  /* extract final carry bit from flags register:
+       x2 <= aa[3072] */
+  csrrs     x2, 0x7c0, x0
+  andi      x2, x2, 1
+
+  /* Compute (aa - M) mod 2^3072:
+       [w16:w27] <= ([w4:w15] - M) = (aa - M) mod 2^3072 */
+  jal       x1, subtract_modulus_var
+
+  /* extract final borrow bit from flags register */
+  csrrs     x3, 0x7c0, x0
+  andi      x3, x3, 1
+
+  /**
+   * Select either aa or aa' based on carry/borrow bits.
+   *
+   * If aa < M, it follows that the carry bit aa[3072] = 0 (since M < 2^3072).
+   * It also follows that the borrow from subtracting M must be 1. In this
+   * case, select aa; otherwise, select aa-M.
+   */
+
+  /* x2 <= (!x2) & x3 */
+  xori      x2, x2, 1
+  and       x2, x2, x3
+
+  /* select aa if x2 = 0, otherwise aa-M */
+  bne       x2, x0, sel_aa
+
+  /* copy subtraction result to w4:w15 */
+  li        x8, 4
+  li        x11, 16
+  loopi     12, 2
+    bn.movr   x8, x11++
+    addi      x8, x8, 1
+
+  sel_aa:
```

Comment:
I think we generally don't indent labels. In non-OTBN asm, we also seem to have tried to be consistent about putting an `L_` prefix on internal lablels: https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/crt/crt.S

Reply:
I might just deindent for now -- existing OTBN assembly doesn't use the `L_` convention as far as I can tell. Maybe it should, but I think it's less confusing to come back change it all at once rather than mix conventions.